File: src/tools/git-status.ts
import { tool } from "ai";
import { runCommand } from "../run-command";
import { z } from "zod";
import { toAgentToolResult } from "../agent-tool-result";
import {
  executeToolWithState,
  type ExecutionTracker,
} from "../execution-state";

type ExecuteGit = (
  args: string[],
) => Promise<{
  stdout: string;
  stderr: string;
  exitCode: number;
}>;

export async function gitStatus(
  executeGit: ExecuteGit = async (args) => {
    return await runCommand("git", args);
  },
) {
  const result = await executeGit(["status", "--short"]);

  return {
    stdout: result.stdout,
    stderr: result.stderr,
    exitCode: result.exitCode,
  };
}

export function createGitStatusTool(options?: {
  executionTracker?: ExecutionTracker;
  executeGit?: ExecuteGit;
}) {
  return tool({
    description: "Show the current git working tree status in short format",

    inputSchema: z.object({}),

    execute: async () => {
      return await toAgentToolResult(async () =>
        await executeToolWithState({
          toolName: "gitStatus",
          tracker: options?.executionTracker,
          run: async () => await gitStatus(options?.executeGit),
        })
      );
    },
  });
}

export const gitStatusTool = createGitStatusTool();

File: internal/tools/gitstatus.go
package tools

import (
	"context"
	"encoding/json"
	"fmt"
)

type gitStatusTool struct{}

func (t *gitStatusTool) Name() string        { return "gitStatus" }
func (t *gitStatusTool) Effect() Effect      { return EffectRead }
func (t *gitStatusTool) Description() string { return "Show the current git working tree status" }
func (t *gitStatusTool) Schema() map[string]any {
	return objectSchema(map[string]any{})
}

func (t *gitStatusTool) Execute(ctx context.Context, input json.RawMessage) (any, error) {
	if err := decodeInput(input, &struct{}{}); err != nil {
		return nil, err
	}
	result, err := runCapturedCommand(ctx, "git", "status", "--porcelain")
	if err != nil {
		return nil, fmt.Errorf("git status failed: %w\n%s", err, result.Stderr)
	}
	return result.Stdout, nil
}

func NewGitStatusTool() Tool { return &gitStatusTool{} }
